util/gen_action_info.py

```diff
@@ -0,0 +1,250 @@
+General streamlining
+Implement column filtering on returned data
+Validate output
+Implement proper logging
+Implement proper error handling
+Review and streamline name normalization (current normalization taken from csv_util.py and coinach.py)
+Build out local file functionality
+Move table structuring from dunders to get_data_table()
+Stab feature creep/overengineering in the eye
+"""
+
+config = {
+    "output": {
+        "pve": "pve_action_info.js",
+        "pvp": "pvp_action_info.js",
+        "crafting": "crafting_action_info.js",
+        "combo": "pve_action_combos.js",
+        "invalid": "invalid_action.log",
+    },
+    "locale_url": {
+        "root": "https://raw.githubusercontent.com/",
+        "intl": "xivapi/ffxiv-datamining/master/csv/",
+        "cn": "thewakingsands/ffxiv-datamining-cn/master/",
+        "ko": "Ra-Workspace/ffxiv-datamining-ko/master/csv/",
+        "local": "",
+    },
+    "path": {"cactbot": os.path.abspath(__file__)[:-24]},
+    "log": {"error": "gen_action_info.log"},
+}
+
+
+def tree():
+    return defaultdict(tree)
+
+
+def __get_remote_table(url, inputs, outputs=None):
+    """Connects to a remote source to retrieve the table data"""
+    # TODO: Error handling
+    with urllib.request.urlopen(url) as response:
+        csv_file = csv.reader(io.StringIO(response.read().decode("utf-8-sig")))
+
+    # First line is the indices and third line is the data types.  They aren't currently used, so we discard them.
+    next(csv_file)
+    # Second line is the headers
+    # Append the hexidecimal version of the ID as a new column so the original ID is usable for correllation
+    headers = next(csv_file) + ["HexID"]
+    next(csv_file)
+
+    # Change # to ID for more readable data
+    headers[0] = headers[0].replace("#", "ID")
+    # Generate the hex ID from the ID and return the data with the headers prepended
+    return [headers] + [x + [format(int(x[0]), "X")] for x in csv_file]
+
+
+def __get_local_table(filename, inputs, outputs=None):
+    """Gets table data from a local file"""
+    # I'm honestly not even sure if this will work as expected. It can be updated/fixed if a use case is found.
+    with open(filename, "r") as table:
+        csv_file = [x.rstrip("\r\n").split(",") for x in table.read().decode("utf-8-sig")]
+
+    # First line is the indices and third line is the data types.  They aren't currently used, so we discard them.
+    csv_file.pop(2)
+    csv_file.pop(0)
+
+    # Next line is the headers
+    # Append the hexidecimal version of the ID as a new column so the original ID is usable for correllation
+    headers = csv_file.pop(0) + ["HexID"]
+    # Change # to ID for more readable data
+    headers[0] = headers[0].replace("#", "ID")
+    # Generate the hex ID from the ID and return the data with the headers prepended
+    return [headers] + [x + [format(int(x[0]), "X")] for x in csv_file]
+
+
+def get_data_table(table_name, locale="intl", inputs=None, outputs=None):
```

Comment:
Is there a reason you can't use `csv_util.get_locale_table` here, or edit it to do what you want?

Reply:
I can definitely update csv_util as well, if that is what is preferred.  There are a lot of specific hard coded things in a lot of this Python that doesn't really make sense to me.  Since I don't know/understand why these are the way they are I wasn't going to presume that I knew better.

For example, there are 4 functions to get a table.  These four functions amount to the same thing, simply with one parameter changed.  Why not simply pass a parameter to one function?

Reply:
I think if there are things that you think can be cleaned up in that Python file, then submit patches to clean them up? I think there's not a strong reason for having four functions that could be one function.  Maybe I thought it was more readable at the time, but I think mostly it just wasn't a big deal either way.  It's a mess, but it's not a huge mess, and so it never got simplified.